Multiplatform support

h-ls opened the pull request:
This PR should fix #36, fix #32 and fix #33

* add support for OS X and Windows platforms
* run unit tests in CI builds
* use Travis for Mac OSX testing: test with XCode 8.3 and 9.4
* use [AppVeyor] for Windows testing: test with Visual Studio 2017
* improve Linux testing: test with GCC 5, 6, 7 and clang 4, 5, 6 (libc++ and stdlibc++) on Ubuntu 14.04
* improve Travis pipeline: deploy docker images and doc only if test stage succeeds
* test find_package(SIRIUS MODULE) and find_package(SIRIUS CONFIG) helpers
* improve CMake scripts
* export `libsirius` and `libsirius-static` targets
* update internal dependencies (catch2, cxxopts, gsl, spdlog)
* refactor resampling and stream API

**Note concerning AppVeyor**: [vcpkg] is used to install FFTW3 and GDAL dependencies.
Given that the build of those dependencies takes a long time (~50min for GDAL and ~3min for FFTW), one build is needed to only generate the cache containing the dependencies. Once the cache is generated, next AppVeyor builds will actually build project and run its tests.
To update the dependencies, delete the cache with [AppVeyor REST API] and run a new build.

[AppVeyor]: https://www.appveyor.com/ "AppVeyor: Continuous Integration solution for Windows and Linux"
[vcpkg]: https://github.com/Microsoft/vcpkg "Package manager for C and C++ libraries on Windows, Linux and MacOS"
[AppVeyor REST API]: https://github.com/appveyor/ci/issues/985

Commit:
#36 run unit tests in Travis CI

Commit:
#33 support MacOS and improve Travis matrix

* test OSX build with XCode 8.3 and 9.4
* test Linux build with GCC5, GCC6 and GCC7 on Ubuntu 14.04
* deploy (doc, docker image) only if tests succeeded

Commit:
#32 support Windows and use AppVeyor for Windows testing

* use vcpkg to build fftw3 and gdal dependencies
* test build with Visual Studio 2017 Win64

Commit:
#36 improve travis matrix

* test Linux build with Clang 4, Clang 5 and Clang 6
* test Clang build using stdlibc++ and libc++

Commit:
minor improvements

* use VERSION option in CMake project function
* use object library to create sirius shared and static libraries
* fix Windows warnings

Commit:
update internal dependencies

* catch2: 2.2.3 -> 2.3.0
* cxxopts: 2.1.0 -> 2.1.1
* gsl: 1.0.0 -> 2.0.0
* spdlog: 0.17.0 -> 1.1.0

Commit:
add configuration header file and test find_package(SIRIUS) helper

configuration definitions:
* SIRIUS_ENABLE_LOGS
* SIRIUS_ENABLE_CACHE_OPTIMIZATION
* GSL_TERMINATE_ON_CONTRACT_VIOLATION
* GSL_UNENFORCED_ON_CONTRACT_VIOLATION

rkanavath commented:
it would be nice to merge this before merging python bindings. for the moment, i copy paste some fixes locally to make build going.

Commit:
improve FFTW3 and SIRIUS find_package helpers

Commit:
add USE_CXX_STATIC_RUNTIME CMake option and rename ENABLE_UNIT_TESTS CMake option to BUILD_TESTING

* test on appveyor
* test on travis (linux - GCC and Clang)

Commit:
update install directory structure

* move runtime artifacts into bin directory (executable, dll)
* move library and archive artifacts into lib directory (so, a, lib)
* update FindSIRIUS.cmake

Commit:
export libsirius and libsirius-static targets at install

enable find_package(SIRIUS CONFIG) with SIRIUS_DIR CMake variable set to /path/INSTALL_DIR/share/cmake

Commit:
add third party license summary and copy licenses into install directory

Commit:
add CHANGELOG.md

Commit:
update library naming
* add debug postfix (d) to libsirius and libsirius-static binary filenames
* rename libsirius-static binary file to libsirius.a (Linux and OS X) or libsirius_s.lib (Windows)

Commit:
update CMakeLists
* group targets by folders (bin, lib, test)
* improve test targets

Commit:
refactor streaming

* generic stream transformation
* update resampling API
* generic image decomposition

(cherry picked from commit 73fb7c69205a6ea7bdf2624b6d5687091eeda183)

Commit:
update catch2 main
* set sirius verbosity level with `-V` option at runtime

Commit:
update README

Commit:
minor fixes

Commit:
appveyor: update dependencies installation

Commit:
appveyor: use vcpkg master branch to install fftw3 and gdal dependencies
